### PR TITLE
[ENH] Implemented IsolatedForest in anomaly detection with some test cases

### LINKED ISSUE #2108: [ENH] Add PyODAdapter-implementation for IsolationForest
### Describe the feature or idea you want to propose

The [`PyODAdapter`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_pyodadapter.py) in aeon allows us to use any outlier detector from [PyOD](https://github.com/yzhao062/pyod), which were originally proposed for relational data, also for time series anomaly detection (TSAD). Not all detectors are equally well suited for TSAD, however. We want to represent the frequently used and competitive outlier detection techniques within the `anomaly_detection` module of aeon directly.

Implement the [IsolationForest method](https://github.com/yzhao062/pyod/blob/master/pyod/models/iforest.py#L49) using the `PyODAdapter`.

### Describe your proposed solution

- Create a new file in `aeon.anomaly_detection` for the method
- Create a new estimator class with `PyODAdapter` as the parent
- Expose the algorithm's hyperparameters as constructor arguments, create the PyOD model and pass it to the super-constructor
- Document your class
- Add tests for certain edge cases if necessary

---

Example for IsolationForest:

```python
class IsolationForest(PyODAdapter):
    """documentation ..."""
    def __init__(n_estimators: int = 100, max_samples: int | str = "auto", ..., window_size: int, stride: int):
        model = IForest(n_estimators, max_samples, ...
        super().__init__(model, window_size, stride)

    @classmethod
    def get_test_params(cls, parameter_set="default"):
        """..."""
        return {"n_estimators": 10, ...}
```

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by notaryanramani**
#### Reference Issues/PRs

Fixes #2108 

#### What does this implement/fix? Explain your changes.

Implements `PyODAdapter` for `IsolationForest` in `anomaly_detection` along with some test cases.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### Any other comments?

Looking for feedback on class documentation & more test cases if needed.

No.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/anomaly_detection/_iforest.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        .. deprecated:: 0.20
+           ``behaviour='old'`` is deprecated in sklearn 0.20 and will not be
+           possible in 0.22.
+
+        .. deprecated:: 0.22
+           ``behaviour`` parameter will be deprecated in sklearn 0.22 and
+           removed in 0.24.
+
+        .. warning::
+            Only applicable for sklearn 0.20 above.
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:75]: dont need these bits, more confusing then useful imo

File: aeon/anomaly_detection/_iforest.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        by `np.random`.
+
+    verbose : int, optional (default=0)
+        Controls the verbosity of the tree building process.
+
+    window_size : int, default=10
+        Size of the sliding window.
+
+    stride : int, default=1
+        Stride of the sliding window.
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:90]: Can you use this style for the above docstrings as well.

File: aeon/anomaly_detection/_iforest.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+from aeon.anomaly_detection._pyodadapter import PyODAdapter
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+class IsolationForest(PyODAdapter):
+    """IForest adapter for anomaly detection.
+
+    This class implements the Isolation Forest algorithm for anomaly detection
+    using PyODAdadpter to be used in the Aeon framework.
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:17]: aeon not Aeon

Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:17]: If the docstrings for parameters are copied from the original, please acknowledge that somewhere here.

Good to mention which parameters are sent directly to the inherited model as well, can just say all parameters buy x, y etc.

File: aeon/anomaly_detection/tests/test_iforest.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        "bootstrap": False,
+        "n_jobs": 1,
+        "behaviour": "old",
+        "random_state": None,
+        "verbose": 0,
+        "window_size": 10,
+        "stride": 1,
+    }
+
+    assert params == expected_dict
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/tests/test_iforest.py:171]: Don't need this test

File: aeon/anomaly_detection/_iforest.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            "contamination": 0.1,
+            "max_features": 1.0,
+            "bootstrap": False,
+            "n_jobs": 1,
+            "behaviour": "old",
+            "random_state": None,
+            "verbose": 0,
+            "window_size": 10,
+            "stride": 1,
+        }
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:178]: These should be the parameters used in unit tests i.e. a reduced value for `n_estimators`. no needs to include defaults.

File: aeon/anomaly_detection/_iforest.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        random_state: Optional[Union[int, np.random.RandomState]] = None,
+        verbose: int = 0,
+        window_size: int = 10,
+        stride: int = 1,
+    ):
+        if not _check_soft_dependencies("pyod", severity="none"):
+            raise ModuleNotFoundError(
+                "pyod is a soft dependency and not included"
+                " in the base aeon installation."
+            )
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_iforest.py:112]: There is an option in `severity` to raise an error, use that instead if checking the dependency is necessary here.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
Few comments. Would be good to hear if you have any additional thoughts @CodeLionX.

File: aeon/anomaly_detection/_iforest.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+class IsolationForest(PyODAdapter):
+    """IForest adapter for anomaly detection.
+
+    This class implements the Isolation Forest algorithm for anomaly detection
+    using PyODAdadpter to be used in the aeon framework. All parameters are passed to
+    IForest except for `window_size` and `stride`.
+
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_iforest.py:20]: Please add the capabilities table to the documentation, similar to the other anomaly detectors. You can copy it from the `PyODAdapter` because the capabilities are the same.

File: aeon/anomaly_detection/_iforest.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        max_features: Union[int, float] = 1.0,
+        bootstrap: bool = False,
+        n_jobs: int = 1,
+        behaviour: str = "old",
+        random_state: Optional[Union[int, np.random.RandomState]] = None,
+        verbose: int = 0,
+        window_size: int = 10,
+        stride: int = 1,
+    ):
+        _check_soft_dependencies("pyod")
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_iforest.py:99]: Please consider using the class-tags instead of this check:

```python
    _tags = {
        "capability:multivariate": True,
        "capability:univariate": True,
        "capability:missing_values": False,
        "fit_is_empty": False,
        "python_dependencies": ["pyod"],
    }
```

File: aeon/anomaly_detection/_iforest.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    def _fit(self, X: np.ndarray, y: Union[np.ndarray, None] = None) -> None:
+        super()._fit(X, y)
+
+    def _predict(self, X: np.ndarray) -> np.ndarray:
+        return super()._predict(X)
+
+    def _fit_predict(
+        self, X: np.ndarray, y: Union[np.ndarray, None] = None
+    ) -> np.ndarray:
+        return super()._fit_predict(X, y)
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_iforest.py:133]: These overrides are not necessary. IMHO, we can remove them.

File: aeon/anomaly_detection/_iforest.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            "contamination": 0.1,
+            "max_features": 1.0,
+            "bootstrap": False,
+            "n_jobs": 1,
+            "behaviour": "old",
+            "random_state": None,
+            "verbose": 0,
+            "window_size": 10,
+            "stride": 1,
+        }
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_iforest.py:165]:
```suggestion
        return {
            "n_estimators": 10,
        }
```

File: aeon/anomaly_detection/tests/test_iforest.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+@pytest.mark.skipif(
+    not _check_soft_dependencies("pyod", severity="none"),
+    reason="required soft dependency PyOD not available",
+)
+def test_iforest_default():
+    """Test IsolationForest."""
+    series = make_series(n_timepoints=80, return_numpy=True, random_state=0)
+    series[50:58] -= 2
+
+    iforest = IsolationForest(window_size=10, stride=1)
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_iforest.py:20]: Please set `random_state` in all tests to avoid spurious test failures.

File: aeon/anomaly_detection/tests/test_iforest.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    pred = iforest.fit_predict(series, axis=0)
+
+    assert pred.shape == (80,)
+    assert pred.dtype == np.float_
+    assert 50 <= np.argmax(pred) <= 60
+    assert hasattr(iforest, "pyod_model")
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_iforest.py:26]: This check is not necessary; also in the following tests.
```suggestion
```

**Review by CodeLionX — CHANGES_REQUESTED** [review]
Thank you for your contributions. I annotated some parts that can be simplified.

File: aeon/anomaly_detection/_iforest.py (hunk 9 above)
Comment (notaryanramani) [reply aeon/anomaly_detection/_iforest.py:133]: I think `_predict` override is necessary because without it, a test in `aeon/testing/estimator_checking/_yield_anomaly_detection_checks.py` fails with the following error:

```bash
        # Test that all anomaly detectors implement abstract predict.
>       assert "_predict" in estimator_class.__dict__
E       AssertionError
```

Or is there another way to implement abstract predict?

Comment (CodeLionX) [reply aeon/anomaly_detection/_iforest.py:133]: Hm, I did not know that we test that 🤔 let me discuss this with the other core devs.

Comment (CodeLionX) [reply aeon/anomaly_detection/_iforest.py:133]: Ok, so we decided to adjust the tests soon. For now, you can keep the current overwrites, and we will remove them once the tests are changed as well.

**Comment by CodeLionX** [comment]
@notaryanramani please do not use force-pushes in the future. This breaks GitHub's differential reviewing feature.

**Review by CodeLionX — CHANGES_REQUESTED** [review]
- Please add the estimator to our API reference in `docs/api_reference/anomaly_detection.rst` (alphabetically sorted).
- We might not need all the tests with different windows because we already test this in the `PyODAdapter` tests. (no change needed)

Otherwise, this looks good, thank you! 👍🏼

**Review by CodeLionX — APPROVED** [review]

**Review by MatthewMiddlehurst — APPROVED** [review]

**Comment by TonyBagnall** [comment]
thanks for this

**Comment by notaryanramani** [comment]
@all-contributors add @notaryanramani for enhancement

**Comment by notaryanramani** [comment]
@all-contributors add @notaryanramani for code
